RE2022-334: add docs re adding new collections (#696)
kbase/collections: docs/genome_to_collection.md

# From KBase Genome to Collection

This guide outlines the process of creating and activating a new collection from KBase narrative with Genome workspace
objects.

> ## Step 1: NERSC Account Setup

Reviewer: ```suggestion
objects.

Note that for all commands you can run `--help` to get more information about the various CLI options.

## Step 1: NERSC Account Setup
```

Reviewer: :+1:

The commands provided in this document are intended for execution within the terminals of 
[NERSC Perlmutter](https://docs.nersc.gov/systems/perlmutter/). 
If you don't have a NERSC account yet, you can create one via [Iris](https://docs.nersc.gov/iris/).
Before proceeding, please ensure that you have an active NERSC account and are logged in to Perlmutter.

```commandline
ssh perlmutter.nersc.gov
```

## Step 2: Prepare Source Data Files (FASTA)

> ### If you already have a KBase narrative with target Genome objects

Reviewer: ```suggestion
## Step 2: Prepare Source Data Files (FASTA)

To run the collections pipeline you will need data uploaded as Genome objects in a KBase Narrative. You can use the Narrative interface to upload data. Additionally, for NCBI data, there are command line based options provided below.

### If you already have a KBase narrative with target Genome objects
```

Reviewer: :+1:

If you already possess a KBase narrative containing the desired Genome objects, download the FASTA files associated with
the Genome objects from the KBase workspace to NERSC. The following command facilitates this process:

```commandline
# update arguments as needed
workspace_id=69739
kbase_collection=ENIGMA
source_verion=2023.10
env=CI
token_filepath=path/to/token_file.txt  # This is a single-line file with the KBase token 

cd /global/cfs/cdirs/kbase/collections/collections
PYTHONPATH=. python src/loaders/workspace_downloader/workspace_downloader.py \
    --workspace_id $workspace_id \
    --kbase_collection $kbase_collection \
    --source_ver $source_verion \

>     --root_dir $root_dir \

Reviewer: This isn't in the list of args above

Author: removed.

Reviewer: :+1:

    --env $env \
    --token_filepath $token_filepath \

>     --retrieve_sample 

Reviewer: Will this error out if there are no or missing samples?

Author: error out. added comments regarding to it.

Reviewer: :+1:

Reviewer: :+1:

```

> ### Download FASTA files from NCBI based on GTDB version

Reviewer: does this allow you to do anything in collections? It seems like it's a dead end

Author: It prepares FASTA files same as what workspace downloader does. We can remove it if we are assuming collections should be created from WS only.

Reviewer: You can run the pipeline and everything works?

Author: oh. They don't have the genome meta data. So parser won't work on them. Remove this section?

Reviewer: If it's a dead end then yeah, probably should

Author: removed this section. I think it makes more sense to just remove it.

Reviewer: :+1:

Alternatively, if you don't possess a KBase narrative, you can download FASTA files from NCBI based on GTDB version 
using the following command:

```commandline
# update arguments as needed
download_file_ext='genomic.fna.gz'
gtdb_release_ver=214

PYTHONPATH=. python src/loaders/ncbi_downloader/gtdb.py \
    --download_file_ext $download_file_ext \
    --release_ver $gtdb_release_ver
```

### Create KBase Genome objects from Genbank files
Optionally, you can also download Genbank files from NCBI and create KBase Genome objects from them using the 
following command. This process will also retrieve the FASTA file from the created Genome objects.

```commandline
# update arguments as needed

# download Genbank files from NCBI based on GTDB version
download_file_ext='genomic.gbff.gz'
gtdb_release_ver=214

PYTHONPATH=. python src/loaders/ncbi_downloader/gtdb.py \
    --download_file_ext $download_file_ext \
    --release_ver $gtdb_release_ver
    
# create KBase Genome objects from the downloaded Genbank files
workspace_id=72231
kbase_collection=GTDB
source_verion=$gtdb_release_ver
load_id=1
au_service_ver=dev

> gfu_service_ver=dev

Reviewer: We should probably leave these out of official docs and let the user discover them via --help if needed.

Author: 👍

Reviewer: :+1:

> cbs_max_tasks=10

Reviewer: Is there any reason not to leave this is as the default?

Author: nope. removed it.

Reviewer: :+1:

PYTHONPATH=. python src/loaders/workspace_uploader/workspace_uploader.py \
    --workspace_id $workspace_id \
    --kbase_collection $kbase_collection \
    --source_ver $source_verion \
    --env $env \
    --token_filepath $token_filepath \
    --au_service_ver $au_service_ver \
    --gfu_service_ver $gfu_service_ver \
    --cbs_max_tasks $cbs_max_tasks \
    --load_id $load_id  
```

# Step 3: Execute tools on FASTA files

Once the source data files are prepared, execute tools on these FASTA files. You can choose from various available 

> tools such as gtdb_tk, checkm2, microtrait, or mash.

Reviewer: Should note here that this will consume NERSC allocation

Author: 👍

Reviewer: :+1:

    --kbase_collection $kbase_collection \
    --source_ver $source_ver \
    --load_ver $load_ver \
    --env $env
	
# Parse genome taxa count information
attri_file=${kbase_collection}_${load_ver}_checkm2_gtdb_tk_kbcoll_genome_attribs.jsonl
PYTHONPATH=. python src/loaders/genome_collection/compute_genome_taxa_count.py \
    ../import_files/$env/$kbase_collection/$load_ver/$attri_file \
    --load_ver $load_ver \
    --kbase_collection $kbase_collection \
    --env $env \
    --input_source genome_attributes
```

## Load parsed results to ArangoDB

```commandline
# set up an SSH tunnel (Not required when using an internal KBase machine such as dev03) 
USER_NAME=user_name                 # user name for login1.berkeley.kbase.us
FORWARD=localhost:48000
ssh -f -N -L $FORWARD:10.58.1.211:8531 \
 $USER_NAME@login1.berkeley.kbase.us

# execute arangoimport (Please ask system admin for the arangoDB credentials)
PARSED_FILE=json_file_path          # the file path generated by the parsing script
ARANGO_USER=arango_username         # arangoDB user name
ARANGO_PW=arango_password           # arangoDB password
ARANGO_DB=collections_dev           # arangoDB database name

> ARANGO_COLL=kbcoll_genome_attribs   # arangoDB collection name

Reviewer: Should explain how to figure out the arango collection for each file and that it needs to be run on all the jsonl files

Author: 👍

Reviewer: :+1:

arangoimport --file $PARSED_FILE \
    --server.endpoint tcp://$FORWARD \
    --server.username $ARANGO_USER \
    --server.password $ARANGO_PW \
    --server.database $ARANGO_DB \
    --collection $ARANGO_COLL \

>     --on-duplicate update

Reviewer: Does this replace the old doc with the new or merge them?

Author: merge operation. added comments.

Reviewer: Hmm, shouldn't it be replace? Why merge?

Author: yea. Now I think about it, it should be `replace`.

Reviewer: Actually... I don't think we should encourage this. If there are differences the load should fail. The whole point of load versions is so you can safely do a new load without worrying about messing with the older loads

Author: 👍 removed `on-duplicate`. By default, it will error.

Reviewer: :+1:

```

> 

Reviewer: I think instructions for the homology files are missing

Author: I actually don't know how homology works. Do you mind adding that section after this PR?

Reviewer: I think you just need to say something along the lines of "for these files, upload them to the homology server using the instructions here https://github.com/jgi-kbase/AssemblyHomologyService and use the namespace when creating the collection with a homology matcher." 

You'll need to pretty that up a bit though, that's just the general idea

Author: 👍

Reviewer: Can you denote which files are the homology files?

Author: 👍 added the sketch file name.

Reviewer: There's 2 files that are product for the homology service, the sketch file and the sequences file

Author: I am guessing you are referring to the sequence metadata file. I included that file.

Reviewer: :+1:

# Step 5: Create and Active the Collection

## 5.1: Verify Permissions
Before proceeding to create and activate the collection, execute the following command to verify if you have the 
required permissions:

```
# Set the server URL and associated token
SERVER='https://ci.kbase.us/services/collections'
TOKEN='your_token_here' 

curl -X 'GET' \
"$SERVER/whoami/" \
-H 'accept: application/json' \
-H "Authorization: Bearer $TOKEN"
```
Example response:
```
{"user":"tgu2","is_service_admin":true}
```
Ensure that the "is_service_admin" field is set to true in the response. If it's false or if you encounter any errors, 
contact system administrator to grant you the necessary permissions.

## 5.2: Create (Save) a New Collection

> Utilize the `Save Collection` endpoint to create a new collection. Execute the following command to create a new collection:

Reviewer: Do we want to direct users to the open api UI rather than, or in addition to, curl?

Author: sure. UI makes more sense.

Reviewer: :+1:

```
COLL_ID='GTDB'
VER_TAG='r207.kbase.2'

> DATA='{"key1": "value1", "key2": "value2"}'

Reviewer: it's not going to be at all clear to a new user what data needs to go in here. The open api UI has a lot of docs around this, maybe reference those

Author: 👍

Reviewer: :+1:

> DATA='{"key1": "value1", "key2": "value2"}'

Reviewer: Also probably needs to explain the relation between load version and data product version in the collections do

Author: I think the open UI has good explanations.

Reviewer: Eh, it's kind of buried in the schema. I'd add a sentence about it

Author: 👍

Reviewer: :+1:

> DATA='{"key1": "value1", "key2": "value2"}'

Reviewer: It's going to be really hard for a brand new person to create this doc from scratch, should probably tell them to look at existing collections for examples

Author: 👍

Reviewer: :+1:

> DATA='{"key1": "value1", "key2": "value2"}'

Reviewer: how do they get search view names?

Author: We should assume that they won't be updating search views here. However, if new users intend to update search views, we should advise them to activate the new collection in the service_manager.py file.

Reviewer: You need the search view names or you can't save the collection

Author: 👍

Reviewer: :+1:

# From KBase Genome to Collection

This guide outlines the process of creating and activating a new collection from KBase narrative with Genome workspace
objects.

Note that for all commands you can run `--help` to get more information about the various CLI options.

## [Step 1: NERSC Account Setup](#step-1-nersc-account-setup)
The commands provided in this document are intended for execution within the terminals of 
[NERSC Perlmutter](https://docs.nersc.gov/systems/perlmutter/). 
If you don't have a NERSC account yet, you can create one via [Iris](https://docs.nersc.gov/iris/).
Before proceeding, please ensure that you have an active NERSC account and are logged in to Perlmutter.

```commandline
ssh perlmutter.nersc.gov
```

## [Step 2: Prepare Source Data Files (FASTA)](#step-2-prepare-source-data-files-fasta)

> To run the collections pipeline you will need data uploaded as Genome objects in a KBase Narrative. You can use the Narrative interface to upload data. Additionally, for NCBI data, there are command line based options provided below.

Reviewer: ```suggestion
To run the collections pipeline you will need data uploaded as Genome objects in a KBase Narrative.
You can use the Narrative interface to upload data. Additionally, for NCBI data, there are command
line based options provided below.
```

The Collections endpoint facilitates the creation and activation of collections. Access to the Collections API is 
provided through the Swagger UI, accessible at the KBase_domain/services/collections/docs URL. For instance, 
you can access the Collections API for the CI environment via the following link: [CI Collections API](https://ci.kbase.us/services/collections/docs).

## [5.1: Verify Permissions](#verify-permissions)

Before proceeding with the creation and activation of the collection, it's essential to confirm that you possess 
the requisite permissions.

To verify your permissions, please begin by authorizing with the KBase Token.

![Authorize](screenshots/authorize.png)

Subsequently, execute the `whoami` endpoint.

![Whoami](screenshots/whoami_endpoint.png)

Example response:
```
{"user":"tgu2","is_service_admin":true}
```
Ensure that the "is_service_admin" field is set to true in the response. If it's false or if you encounter any errors, 
contact system administrator to grant you the necessary permissions.

## [5.2: Create (Save) a New Collection](#create-save-a-new-collection)
The next step is to create a new collection with new or updated data.

> ### [5.2.1: Retrieve existing collection data](#retrieve-existing-collection-data)
> If you're updating an existing collection or uncertain about collection's data structure, you can retrieve the data 
> from the currently activated collection by using the `get_collection` endpoint.

Reviewer: ```suggestion
### [5.2.1: Retrieve existing collection data](#retrieve-existing-collection-data)
If you're updating an existing collection or creating a new collection and need an example data structure,
you can retrieve the data from a currently activated collection by using the `get_collection` endpoint.
```
